src/strands/event_loop/event_loop.py

```diff
@@ -372,12 +372,20 @@ async def _handle_model_execution(
                 if model_invoke_span:
                     tracer.end_span_with_error(model_invoke_span, str(e), e)
 
-                await agent.hooks.invoke_callbacks_async(
-                    AfterModelCallEvent(
-                        agent=agent,
-                        exception=e,
-                    )
+                after_model_call_event = AfterModelCallEvent(
+                    agent=agent,
+                    exception=e,
                 )
+                await agent.hooks.invoke_callbacks_async(after_model_call_event)
+
+                # Check if hooks want to retry the model call
+                if after_model_call_event.retry_model and after_model_call_event.exception:
```

Comment:
Go through the exercise of adding support for retrying the the model call even when there aren't exceptions.  Add appropriate plumbing and test cases for that scenario